zcash_client_backend/src/data_api/testing/pool.rs

```diff
@@ -1,36 +1,54 @@
 use assert_matches::assert_matches;
-use incrementalmerkletree::Level;
+use incrementalmerkletree::{frontier::Frontier, Level};
 use rand::RngCore;
+use secrecy::Secret;
 use shardtree::error::ShardTreeError;
-use std::{cmp::Eq, convert::Infallible, hash::Hash, num::NonZeroU32};
+use std::{
+    cmp::Eq,
+    convert::Infallible,
+    hash::Hash,
+    num::{NonZeroU32, NonZeroU8},
+};
+use zip32::Scope;
 
 use zcash_keys::{address::Address, keys::UnifiedSpendingKey};
 use zcash_primitives::{
     block::BlockHash,
-    transaction::{fees::StandardFeeRule, Transaction},
+    legacy::TransparentAddress,
+    transaction::{
+        components::amount::NonNegativeAmount,
+        fees::{fixed::FeeRule as FixedFeeRule, StandardFeeRule},
+        Transaction,
+    },
 };
 use zcash_protocol::{
-    consensus::{self, BlockHeight},
+    consensus::{self, BlockHeight, NetworkUpgrade, Parameters},
     memo::{Memo, MemoBytes},
     value::Zatoshis,
     ShieldedProtocol,
 };
-use zip321::Payment;
+use zip321::{Payment, TransactionRequest};
 
 use crate::{
     data_api::{
-        chain::{CommitmentTreeRoot, ScanSummary},
-        testing::{AddressType, TestBuilder},
+        self,
+        chain::{ChainState, CommitmentTreeRoot, ScanSummary},
+        testing::{input_selector, AddressType, FakeCompactOutput, InitialChainState, TestBuilder},
         wallet::{decrypt_and_store_transaction, input_selection::GreedyInputSelector},
-        Account as _, DecryptedTransaction, InputSource, WalletCommitmentTrees, WalletRead,
-        WalletSummary,
+        Account as _, AccountBirthday, DecryptedTransaction, InputSource, Ratio,
+        WalletCommitmentTrees, WalletRead, WalletSummary, WalletWrite,
     },
     decrypt_transaction,
-    fees::{standard, DustOutputPolicy},
+    fees::{fixed::SingleOutputChangeStrategy, standard, DustOutputPolicy},
     wallet::{Note, NoteId, OvkPolicy, ReceivedNote},
 };
 
-use super::{DataStoreFactory, TestCache, TestFvk, TestState};
+use super::{DataStoreFactory, Reset, TestCache, TestFvk, TestState};
+
+/// The number of ephemeral addresses that can be safely reserved without observing any
+/// of them to be mined. This is the same as the gap limit in Bitcoin.
+#[cfg(feature = "transparent-inputs")]
+pub(crate) const GAP_LIMIT: u32 = 20;
```

Comment:
Does this belong elsewhere maybe?

Reply:
Yes; it can't live under `zcash_client_backend::data_api_testing` as that's only accessible with the `test-dependencies` feature flag.